library/include/modules/Persistence.h

```diff
@@ -0,0 +1,232 @@
+        uint32_t get_uint28(size_t off) const
+        {
+            auto p = pdata<int28_size>(off);
+            return (p[0]>>1) | ((p[1]&~1U)<<6) | ((p[2]&~1U)<<13) | ((p[3]&~1U)<<20);
+        }
+        int32_t get_int28(size_t off) const
+        {
+            return int32_t(get_uint28(off));
+        }
+        void set_uint28(size_t off, uint32_t val)
+        {
+            auto p = pdata<int28_size>(off);
+            p[0] = uint8_t((val<<1) | 1);
+            p[1] = uint8_t((val>>6) | 1);
+            p[2] = uint8_t((val>>13) | 1);
+            p[3] = uint8_t((val>>20) | 1);
+        }
+        void set_int28(size_t off, int32_t val)
+        {
+            set_uint28(off, val);
+        }
+
+        PersistentDataItem() : index(0), data(nullptr) {}
+        PersistentDataItem(size_t index, const std::shared_ptr<Persistence::LegacyData> &data)
+            : index(index), data(data) {}
+    };
+    namespace Persistence
+    {
+        class Internal
+        {
+            static void clear();
+            static void save();
+            static void load();
+            friend class ::DFHack::Core;
+        };
+
+        // Returns a new PersistentDataItem with the specified key.
+        // If there is no world loaded or the key is empty, returns an invalid item.
+        DFHACK_EXPORT PersistentDataItem addItem(const std::string &key);
+        // Returns an existing PersistentDataItem with the specified key.
+        // If "added" is not null and there is no such item, a new item is returned and
+        // the bool value is set to true. If "added" is not null and an item is found or
+        // no new item can be created, the bool value is set to false.
+        DFHACK_EXPORT PersistentDataItem getByKey(const std::string &key, bool *added = nullptr);
+        // Returns an existing PersistentDataItem with the specified index.
+        // If there is no world loaded or the index is empty, returns an invalid item.
+        DFHACK_EXPORT PersistentDataItem getByIndex(size_t index);
+        // If the item is invalid, returns false. Otherwise, deletes the item and returns
+        // true. All references to the item are invalid as soon as this function returns.
+        DFHACK_EXPORT bool deleteItem(const PersistentDataItem &item);
+        // Fills the vector with references to each persistent item.
+        DFHACK_EXPORT void getAll(std::vector<PersistentDataItem> &vec);
+        // Fills the vector with references to each persistent item with a key that is
+        // greater than or equal to "min" and less than "max".
+        DFHACK_EXPORT void getAllByKeyRange(std::vector<PersistentDataItem> &vec, const std::string &min, const std::string &max);
+        // Fills the vector with references to each persistent item with a key that is
+        // equal to the given key.
+        DFHACK_EXPORT void getAllByKey(std::vector<PersistentDataItem> &vec, const std::string &key);
+
+#if defined(__GNUC__) && __GNUC__ < 5
```

Comment:
Perhaps this could go in a utility file - I've seen a compiler workarounds header, but probably in an unmerged PR.